### PR TITLE
Changes for feature environments

### PR DESCRIPTION
**Pull request by namanmishra91**
- Added namespacing to producer/consumer and queues
- Set TTL for queues

### CODE REVIEW

File: src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java [hunk 1]
```diff
@@ -1,16 +1,28 @@
...
-        return String.format("%s.%s", prefix, name);
+        String namespaced = prefixWithNamespace(name);
+        return String.format("%s.%s", prefix, namespaced);
     }
 
     public static String sanitizeMetricName(String metric) {
         return metric == null ? null : metric.replaceAll("[^A-Za-z\\-0-9]", "").toLowerCase();
     }
+
+    public static String prefixWithNamespace(String name) {
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java:21]: `val`

File: src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java [hunk 2]
```diff
@@ -163,6 +164,33 @@ public void ensure(final String queueName,
                 .build();
     }
 
+    public Map<String, Object> rmqOpts(boolean enableQueueTTL, int ttl) {
+        Map<String, Object> ttlOpts = getTTLOpts(enableQueueTTL, ttl);
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java:168]: final/val .. do this in other places. Not commenting on same issue anymore

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/dropwizard/actors/actor/ActorConfig.java [hunk 3]
```diff
@@ -67,6 +67,13 @@
     @Builder.Default
     private RetryConfig retryConfig = new NoRetryConfig();
 
+    @Builder.Default
+    private boolean queueTTLEnabled = false;
+
+    // TTL in milliseconds
+    @Builder.Default
+    private int ttl = 1800000;
```
[the remark this answers is not in the document]
Comment (namanmishra91) [reply src/main/java/io/appform/dropwizard/actors/actor/ActorConfig.java]: Done

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 4]
```diff
@@ -87,10 +87,10 @@ public void start() throws Exception {
         ensureExchange(dlx);
 
         this.publishChannel = connection.newChannel();
-        connection.ensure(queueName + "_SIDELINE", queueName, dlx);
-        connection.ensure(queueName, config.getExchange(), connection.rmqOpts(dlx));
+        connection.ensure(queueName + "_SIDELINE", queueName, dlx, connection.rmqOpts(config.isQueueTTLEnabled(), config.getTtl()));
+        connection.ensure(queueName, config.getExchange(), connection.rmqOpts(dlx, config.isQueueTTLEnabled(), config.getTtl()));
         if (config.getDelayType() == DelayType.TTL) {
-            connection.ensure(ttlQueue(queueName), queueName, ttlExchange(config), connection.rmqOpts(exchange));
+            connection.ensure(ttlQueue(queueName), queueName, ttlExchange(config), connection.rmqOpts(exchange, config.isQueueTTLEnabled(), config.getTtl()));
```
[the remark this answers is not in the document]
Comment (namanmishra91) [reply src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java]: Done

File: src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java [hunk 5]
```diff
@@ -1,16 +1,28 @@
 package io.appform.dropwizard.actors.base.utils;
 
+import io.appform.dropwizard.actors.utils.CommonUtils;
 import lombok.AccessLevel;
 import lombok.NoArgsConstructor;
 
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 public class NamingUtils {
 
+    public static final String FEATURE_ENV_NAME = "feature_env_name";
```
[the remark this answers is not in the document]
Comment (namanmishra91) [reply src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java]: Done

File: src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java [hunk 6]
```diff
@@ -1,16 +1,28 @@
...
+
     public static String queueName(String prefix, String name) {
-        return String.format("%s.%s", prefix, name);
+        String namespaced = prefixWithNamespace(name);
```
[the remark this answers is not in the document]
Comment (namanmishra91) [reply src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java]: Changed local variables to `final`

**Comment by namanmishra91** [comment]
@r0goyal Can you please review this as well?

File: src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java (hunk 1 above)
Comment (namanmishra91) [reply src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java:21]: @santanusinha Changed the local variables to final. Did you meant the same for function argument as well?

File: pom.xml [hunk 7]
```diff
@@ -68,6 +68,7 @@
         <lombok.version>1.18.8</lombok.version>
         <guava.version>28.0-jre</guava.version>
         <junit.version>4.12</junit.version>
+        <test.container.version>1.0.4</test.container.version>
```
Comment (santanusinha) [inline pom.xml:71]: Please add yourself to developer section

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: src/test/java/io/appform/dropwizard/actors/connectivity/actor/NamespacedQueuesTest.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        if (response != null) {
+            JsonNode jsonNode = objectMapper.readTree(response.body().string());
+            if (jsonNode.isArray()) {
+                for (JsonNode json: jsonNode) {
+                    String queueName = json.get("name").asText();
+                    Assert.assertTrue(queueName.contains(FEATURE_ENV_NAME));
+                }
+            }
+        }
+        response.close();
```
Comment (namanmishra91) [inline src/test/java/io/appform/dropwizard/actors/connectivity/actor/NamespacedQueuesTest.java:85]: This needs to be moved inside `if(response != null)`. Will do

**Comment by r0goyal** [comment]
@santanusinha @namanmishra91 Currently ttl enabled is scoped to individual actors.. Shall we scope it to the bundle itself - Essentially either all actors would be ttl based or none of them would be ttl based ?

File: src/main/java/io/appform/dropwizard/actors/actor/ActorConfig.java [hunk 9]
```diff
@@ -67,6 +68,13 @@
     @Builder.Default
     private RetryConfig retryConfig = new NoRetryConfig();
 
+    @Builder.Default
+    private boolean queueTTLEnabled = false;
+
+    // TTL in seconds
+    @Builder.Default
+    private Duration ttl = Duration.ofSeconds(1800);
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/actors/actor/ActorConfig.java:76]: @santanusinha Shall we just keep a default value for this - Say 30 seconds or a minute and not let this be overridden

File: src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java [hunk 10]
```diff
@@ -1,16 +1,28 @@
 package io.appform.dropwizard.actors.base.utils;
 
+import io.appform.dropwizard.actors.utils.CommonUtils;
 import lombok.AccessLevel;
 import lombok.NoArgsConstructor;
 
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 public class NamingUtils {
 
+    public static final String FEATURE_ENV_NAME = "FEATURE_ENV_NAME";
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/actors/base/utils/NamingUtils.java:10]: Rename this to NAMESPACE_ENV_NAME

**Comment by namanmishra91** [comment]
> @santanusinha @namanmishra91 Currently ttl enabled is scoped to individual actors.. Shall we scope it to the bundle itself - Essentially either all actors would be ttl based or none of them would be ttl based ?
Was taken care of by @r0goyal

File: src/test/java/io/appform/dropwizard/actors/connectivity/actor/NamespacedQueuesTest.java (hunk 8 above)
Comment (namanmishra91) [reply src/test/java/io/appform/dropwizard/actors/connectivity/actor/NamespacedQueuesTest.java:86]: Done

**Comment by namanmishra91** [comment]
@santanusinha PR is ready for review after changes

File: src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java [hunk 11]
```diff
@@ -148,21 +147,41 @@ public void ensure(final String queueName,
...
+        final Map<String, Object> ttlOpts = getActorTTLOpts(actorConfig.getTtlConfig());
         return ImmutableMap.<String, Object>builder()
+                .putAll(ttlOpts)
                 .put("x-ha-policy", "all")
                 .put("ha-mode", "all")
                 .put("x-dead-letter-exchange", deadLetterExchange)
                 .build();
     }
 
+    private Map<String, Object> getActorTTLOpts(final TtlConfig ttlConfig) {
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java:170]: Move all private functions to after public ones

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java (hunk 11 above)
Comment (namanmishra91) [reply src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java]: Done

**Review by santanusinha — APPROVED** [review]

**Pull request merged**